Fix variable assignment in try-with-resources pattern

sadikovi opened the pull request:
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change
This PR fixes issue with broken highlighting when resource expression is not closed with `;` in try-with-resources pattern, which is technically allowed. @50Wliu suggested to come up with separate pattern for try-with-resources, which I simplified a bit by keeping original try-catch-finally, but instead creating separate pattern for variable definition. 

Such pattern is called `variables-definition-only` and used in try-with-resources to match the following: `String variable =` or `int num =` without actual expression. We use `#code` in patterns to highlight the expression code. 

In this case `meta.definition.variable.java` only covers type, variable name and `=`. I am not sure if this is required to add code, but it highlights correctly.

After patch (also passes test case mentioned in the issue):
<img width="294" alt="after-patch" src="https://user-images.githubusercontent.com/7788766/35179953-3043e56e-fe0a-11e7-8757-98f06bc0ddf2.png">

### Alternate Designs
Alternative design was considered where we apply a hack to `end` by listing `)` as another terminating character. This is a very small fix (4-8 characters change), but could potentially result in other issues, such as triggering close on closing paren (see https://github.com/atom/language-java/issues/88#issuecomment-333003571).

### Benefits
Fixes issue where not using `;` (which is allowed by specification) would break the highlighting. This change is relatively minor and only affects `try-with-resources`.

### Possible Drawbacks
Since we have to duplicate large regular expression to capture variables (in `variables` and `variables-definition-only`), this could potentially result in updating only one but not the other. I added comment that both patterns should be updated and be in sync, but this might be enough.

### Applicable Issues
Fixes #88

sadikovi commented:
@50Wliu could you review this PR, please? Thanks!

50Wliu commented:
Will look at both of your PRs next week hopefully.

50Wliu commented on grammars/java.cson:
```diff
@@ -544,6 +544,9 @@
+                'include': '#variables-definition-only'
+              }
               {
                 'include': '#code'
```
I don't think this is needed anymore.  Try-with-resources only supports variable definitions.

sadikovi replied:
I tried removing `#code`, but it affects syntax highlighting after `=`, for example, in `try (Buffer a = new Buffer())`, `new Buffer()` part will not be highlighted. So I decided to keep it.

50Wliu commented on grammars/java.cson:
```diff
@@ -1203,6 +1206,53 @@
       }
     ]
+  # pattern for variable name
+  'variables-name':
```
-> `variable-name`

sadikovi replied:
👍

50Wliu commented on grammars/java.cson:
```diff
@@ -1203,6 +1206,53 @@
+  # pattern for definition of a variable including assignment, but without expression part, e.g.
+  # 'String variable =' or 'int num =', used with '#code' when we cannot rely on ';' to terminate
+  # variable assignment
+  'variables-definition-only':
```
What do you think about `variable-assignment` or something?

sadikovi replied:
👍

50Wliu commented on grammars/java.cson:
```diff
@@ -1203,6 +1206,53 @@
+        'include': '#all-types'
+      }
+      {
+        'match': '='
```
Have you considered making this a begin/end like `variables` but ending it on `(?!\\G)` instead?  I'm curious if that will work or not.

sadikovi replied:
I have not considered it, but I updated the code and it works (assuming that I updated it with what you had in mind).

sadikovi commented:
@50Wliu I addressed your comments. One thing that I decided not to do is removing `#code` from try with resources, because it does not highlight part after `=`. But it could also mean that I missed something - would appreciate if you could review again.

Also I changed match to begin/end, and it worked. Could you review that part to make sure that I did it right? I do not quite understand the `?!\\G` pattern.

Thanks!

50Wliu commented:
`\G` means "where the last match ended".  So `\Ga` means "match `a` only if it immediately follows the last match".  Therefore, its negation `(?!\G)` means something akin to "start matching only if we are not immediately following where the previous match ended".  So doing `'end': '(?!\\G)'` then means "end the match when there are no more continuous matches".

You'll have to test something like `var = 3 + 2` to make sure it works correctly though, because I still have trouble deciding when to use `\G`.

sadikovi commented on grammars/java.cson:
```diff
@@ -1203,6 +1206,53 @@
+        'beginCaptures':
+          '0':
+            'name': 'keyword.operator.assignment.java'
+        'end': '(?!\\G)'
```
I tried it, updated code as follows:
```diff
diff --git a/grammars/java.cson b/grammars/java.cson
index c6c513e..6700101 100644
--- a/grammars/java.cson
+++ b/grammars/java.cson
@@ -547,9 +547,6 @@
               {
                 'include': '#variable-assignment'
               }
-              {
-                'include': '#code'
-              }
             ]
           }
           {
@@ -1249,6 +1246,11 @@
           '0':
             'name': 'keyword.operator.assignment.java'
         'end': '(?!\\G)'
+        'patterns': [
+          {
+            'include': '#code'
+          }
+        ]
       }
     ]
   # pattern for full variable definition
```

Did not work. I also tried couple of different combinations - same result.

sadikovi commented:
@50Wliu I updated the code. Have you decided if we should go with `\G`? What do you think about overall approach? Do you want me to make any other changes? Thanks.

sadikovi commented:
@50Wliu Is this change still relevant? If so, I suggest we deal with other opened PRs, so I can rebase this one and update if necessary (because of some changes in other pull requests).

50Wliu commented:
Since this change is a little larger I'll be reviewing it separately within the coming days.

sadikovi commented:
@50Wliu I rebased this PR. Could you review it again, when you have time? Thanks!
Do you have any concerns with this work? One thing I can think of is duplicated code for `variable-assignment` and `variables`, but I do not know how to make it reusable.

sadikovi commented:
This needs additional reviewing and potentially different fix, since we have to duplicate patterns here. I will have a look again, see if there is a better approach.

sadikovi commented:
I am going to close this PR, since it introduces a lot of unnecessary changes. I will open an alternative solution shortly, so we can choose the one to go with.
